File: prediction-job/src/main/scala/org/fiware/cosmos/orion/spark/connector/prediction/TrainingJob.scala
package org.fiware.cosmos.orion.spark.connector.prediction

import org.apache.spark.ml.Pipeline
import org.apache.spark.ml.evaluation.RegressionEvaluator
import org.apache.spark.ml.feature.{VectorAssembler, VectorIndexer, StringIndexer, OneHotEncoder}
import org.apache.spark.ml.classification._
import org.apache.spark.sql.{DataFrame, SparkSession}
import org.apache.spark.sql.types.{IntegerType, StringType, DoubleType, StructField, StructType}

object TrainingJob {

  def main(args: Array[String]): Unit = {
    train().write.overwrite().save("./prediction-job/model")

  }
  
  def train() = {
    // ocupation= round((1-available)*10)
    val schema = StructType(
      Array(StructField("name", StringType),
            StructField("availableSpotNumber", IntegerType),
            StructField("timestamp", StringType),
            StructField("weekday", IntegerType),
            StructField("total", IntegerType),
            StructField("day", IntegerType),
            StructField("month", IntegerType),
            StructField("hour", IntegerType),
            StructField("minute", IntegerType),
            StructField("hour_interval", IntegerType),
            StructField("time", StringType),  
            StructField("available", DoubleType), 
            StructField("occupation", IntegerType)                   
      ))
    val spark = SparkSession
      .builder
      .appName("TrainingParkingMalaga")
      .master("local[*]")
      .getOrCreate()

    import spark.implicits._
    spark.sparkContext.setLogLevel("WARN")

    // convert to dataframe
    val data = spark.read.format("csv")
      .schema(schema)
      .option("header", "true")
      .option("delimiter", ",")
      .load("./prediction-job/malaga_parking.csv")

    // onehotencoder
    val it = Array("name", "weekday")
    val stringIndexers = it.map (
        c => new StringIndexer().setInputCol(c).setOutputCol(s"${c}-index")
    )
    val oneHotEncoders = it.map (
        c => new OneHotEncoder().setInputCol(s"${c}-index").setOutputCol(s"${c}-ohe")
    )
              
    val vectorAssembler  = new VectorAssembler()
      .setInputCols(Array("name-ohe","weekday-ohe","hour","month"))
      .setOutputCol("features")

    // Automatically identify categorical features, and index them.
    // var transformedDf = vectorAssembler.transform(data)

    //val Array(trainingData, testData) = transformedDf.randomSplit(Array(0.8, 0.2))


    val rfc = new RandomForestClassifier()
      .setNumTrees(200)
      .setFeatureSubsetStrategy("log2")
      .setLabelCol("occupation")
      .setFeaturesCol("features")

    val pipeline = new Pipeline().setStages(stringIndexers ++ oneHotEncoders ++ Array(vectorAssembler,rfc))
    val Array(trainingData,testData) = data.randomSplit(Array(0.8,0.2))
    val model = pipeline.fit(trainingData)
    val predictions = model.transform(testData)

    predictions.select("prediction","occupation", "hour", "weekday", "name","month").show(10)

    // val evaluator = new MulticlassClassificationEvaluator()
    //   .setLabelCol("occupation")
    //   .setPredictionCol("prediction")
    //   .setMetricName("accuracy")
    // val accuracy = evaluator.evaluate(predictions)
    // println(s"Accuracy = ${accuracy}")

    //println(s"Learned classification forest model:\n ${rfModel.toDebugString}")

    //return the model with the pipeline
    model
  }
}

File: prediction-job/TrainingJob.py
from pyspark.ml import Pipeline
from pyspark.ml.evaluation import RegressionEvaluator
from pyspark.ml.feature import VectorAssembler, VectorIndexer, StringIndexer, OneHotEncoder
from pyspark.ml.classification import RandomForestClassifier
from pyspark.sql import SparkSession
from pyspark.sql.types import IntegerType, StringType, StructField, StructType, DoubleType
import mlflow
from mlflow.entities import RunStatus
from mlflow import MlflowClient
import mlflow.tracking
import mlflow.spark
import os
import datetime
import multiprocessing

MASTER = os.environ.get("SPARK_MASTER", "local[*]")
#MASTER = "local[*]"
MLFLOW_HOST = os.environ.get("MLFLOW_HOST", "localhost")
CSV_FILE = os.environ.get("CSV_FILE", "malaga_parking.csv")
MLFLOW_URI = "http://"+MLFLOW_HOST+":5000"
BASE_PATH = os.environ.get("PROYECT_BASE_PATH", "/prediction-job")

def train():
    schema = StructType([
        StructField("name", StringType()),
        StructField("availableSpotNumber", IntegerType()),
        StructField("timestamp", StringType()),
        StructField("weekday", IntegerType()),
        StructField("total", IntegerType()),
        StructField("day", IntegerType()),
        StructField("month", IntegerType()),
        StructField("hour", IntegerType()),
        StructField("minute", IntegerType()),
        StructField("hour_interval", IntegerType()),
        StructField("time", StringType()),
        StructField("available", DoubleType()),
        StructField("occupation", IntegerType())
    ])

    # Initialize MLFlow connection and experiment

    
    mlflow.set_experiment_tag("createdAt", datetime.datetime.now().strftime("%Y-%m-%d_%H-%M-%S-%f"))

    spark = SparkSession.builder \
        .appName("TrainingJob") \
        .master(MASTER) \
        .config("spark.cores.max", str(int(multiprocessing.cpu_count() / 2))) \
        .getOrCreate()

    spark.sparkContext.setLogLevel("WARN")

    # Load and parse the data file, converting it to a DataFrame.
    data = spark.read.format("csv") \
        .schema(schema) \
        .option("header", "true") \
        .option("delimiter", ",") \
        .load(BASE_PATH+"/"+CSV_FILE)

    # Define columns for one-hot encoding
    categorical_cols = ["name", "weekday"]

    string_indexers = [StringIndexer(inputCol=col, outputCol=col + "-index") for col in categorical_cols]

    one_hot_encoders = [OneHotEncoder(inputCol=col + "-index", outputCol=col + "-ohe") for col in categorical_cols]

    feature_cols = ["name-ohe", "weekday-ohe", "hour", "month"]
    vector_assembler = VectorAssembler(inputCols=feature_cols, outputCol="features")


    num_trees = 200

    mlflow.log_param("numTrees", str(num_trees))

    # Train a RandomForest model.
    rfc = RandomForestClassifier(numTrees=num_trees, featureSubsetStrategy="log2", labelCol="occupation", featuresCol="features")

    stages = string_indexers + one_hot_encoders + [vector_assembler, rfc]
    pipeline = Pipeline(stages=stages)

    # Split the data into training and testing sets
    trainingData, testData = data.randomSplit([0.8, 0.2])

    model = pipeline.fit(trainingData)
    
    model_path = BASE_PATH+"/model"
    rf_model = model.stages[1]

    # Register model
    model_name = 'parkingModel'
    mlflow.spark.log_model(spark_model=model, artifact_path=model_name, registered_model_name=model_name)

    client = MlflowClient(tracking_uri=MLFLOW_URI)
    model_version = client.get_latest_versions(model_name, stages=["None"])[0].version
    model.save(model_path+"/"+model_version)



if __name__ == "__main__":
    mlflow.set_tracking_uri(MLFLOW_URI)
    mlflow.set_experiment("Experiment")
    experiment = mlflow.get_experiment_by_name("Experiment")
    experiment_id = experiment.experiment_id
    with mlflow.start_run(experiment_id=experiment_id, description="Model for predicting parking state") as run:
        train()